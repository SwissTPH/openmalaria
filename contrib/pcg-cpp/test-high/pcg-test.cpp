/*
 * PCG Random Number Generation for C++
 *
 * Copyright 2014-2017 Melissa O'Neill <oneill@pcg-random.org>,
 *                     and the PCG Project contributors.
 *
 * SPDX-License-Identifier: (Apache-2.0 OR MIT)
 *
 * Licensed under the Apache License, Version 2.0 (provided in
 * LICENSE-APACHE.txt and at http://www.apache.org/licenses/LICENSE-2.0)
 * or under the MIT license (provided in LICENSE-MIT.txt and at
 * http://opensource.org/licenses/MIT), at your option. This file may not
 * be copied, modified, or distributed except according to those terms.
 *
 * Distributed on an "AS IS" BASIS, WITHOUT WARRANTY OF ANY KIND, either
 * express or implied.  See your chosen license for details.
 *
 * For additional information about the PCG random number generation scheme,
 * visit http://www.pcg-random.org/.
 */

/*
 * This file is based on the demo program for the C generation schemes.
 * It shows some basic generation tasks.
 */

#include <cstddef>
#include <cstdint>
#include <cstdlib>
#include <cassert>
#include <climits>
#include <iostream>
#include <sstream>
#include <iomanip>
#include <algorithm>
#include <numeric>
#include <random>       // for random_device

#include "pcg_random.hpp"

// This code can be compiled with the preprocessor symbol RNG set to the
// PCG generator you'd like to test.

#ifndef RNG
    #define RNG pcg32
    #define TWO_ARG_INIT 1
#endif

#define STRINGIFY_IMPL(x) #x
#define STRINGIFY(x)      STRINGIFY_IMPL(x)

using namespace std;
using pcg_extras::operator<<;

#if !PCG_EMULATED_128BIT_MATH || !AWKWARD_128BIT_CODE

int main(int argc, char** argv)
{
    // Read command-line options

    int rounds = 5;
    bool nondeterministic_seed = false;

    ++argv;
    --argc;
    if (argc > 0 && strcmp(argv[0], "-r") == 0) {
        nondeterministic_seed = true;
        ++argv;
        --argc;
    }
    if (argc > 0) {
        rounds = atoi(argv[0]);
    }

    /* Many of the generators can be initialized with two arguments; the second
     * one specifies the stream.
     */

#if TWO_ARG_INIT
    RNG rng(42u, 54u);
#else
    RNG rng(42u);
#endif

    if (nondeterministic_seed) {
        // Seed with external entropy from std::random_device (a handy
        // utility provided by pcg_extras).
        rng.seed(pcg_extras::seed_seq_from<std::random_device>());
    }

    constexpr auto bits = sizeof(RNG::result_type) * CHAR_BIT;
    constexpr int how_many_nums = bits <= 8  ? 14
                                : bits <= 16 ? 10
                                :              6;
    constexpr int wrap_nums_at  = bits >  64 ? 2
                                : bits >  32 ? 3
                                :              how_many_nums;

    cout << STRINGIFY(RNG) << ":\n"
    //   << "      -  aka:         " << pcg_extras::printable_typename<RNG>()
    // ^-- we skip this line because the output is long, scary, ugly, and
    //     and varies depending on the platform
         << "      -  result:      " <<  bits << "-bit unsigned int\n"
         << "      -  period:      2^" << RNG::period_pow2();
    if (RNG::streams_pow2() > 0)
         cout << "   (* 2^" << RNG::streams_pow2() << " streams)";
    cout << "\n      -  size:        " << sizeof(RNG) << " bytes\n\n";

    for (int round = 1; round <= rounds; ++round) {
        printf("Round %d:\n", round);

        /* Make some N-bit numbers */
        cout << setw(4) << setfill(' ') << bits << "bit:";
        for (int i = 0; i < how_many_nums; ++i) {
            if (i > 0 && i % wrap_nums_at == 0)
                cout << "\n\t";
            cout << " 0x" << hex << setfill('0') 
                 << setw(sizeof(RNG::result_type)*2) << rng();
        }
        cout << endl;

        cout << "  Again:";
        rng.backstep(6);
        for (int i = 0; i < how_many_nums; ++i) {
            if (i > 0 && i % wrap_nums_at == 0)
                cout << "\n\t";
            cout << " 0x" << hex << setfill('0') 
                 << setw(sizeof(RNG::result_type)*2) << rng();
        }
        cout << dec << endl;

        /* Toss some coins */
        cout << "  Coins: ";
        for (int i = 0; i < 65; ++i)
            cout << (rng(2) ? "H" : "T");
        cout << endl;

        RNG rng_copy{rng};
        /* Roll some dice */
        printf("  Rolls:");
        for (int i = 0; i < 33; ++i)
            cout << " " << (uint32_t(rng(6)) + 1);
        cout << "\n   -->   rolling dice used "
             << static_cast<size_t>(rng - rng_copy)
             << " random numbers" << endl;

        /* Deal some cards using pcg_extras::shuffle, which follows
         * the algorithm for shuffling that most programmers would expect.
         * (It's unspecified how std::shuffle works.)
         */      
        enum { SUITS = 4, NUMBERS = 13, CARDS = 52 };
        char cards[CARDS];
        iota(begin(cards), end(cards), 0);
        pcg_extras::shuffle(begin(cards), end(cards), rng);

        /* Output the shuffled deck */
        printf("  Cards:");
        static const signed char number[] = {'A', '2', '3', '4', '5', '6', '7',
                                             '8', '9', 'T', 'J', 'Q', 'K'};
        static const signed char suit[] =   {'h', 'c', 'd', 's'};
        int i = 0;
        for (auto card : cards) {
            ++i;
            cout << " " << number[card / SUITS] << suit[card % SUITS];
            if (i % 22 == 0)
                cout << "\n\t";
        }
        
        cout << "\n" << endl;
    }

    // Ensure that the input/output functions work.
    std::stringstream ss;
    ss << rng;
    RNG from_string;
    ss >> from_string;
    assert(from_string == rng);

    return 0;
}

#else //  i.e. PCG_EMULATED_128BIT_MATH && AWKWARD_128BIT_CODE

int main()
{
    // It's not that it *can't* be done, it just requires either C++14-style
    // constexpr or some things not to be declared const.
    cout << "Sorry, " STRINGIFY(RNG) " not supported with emulated 128-bit math"
         << endl;
}

#endif
