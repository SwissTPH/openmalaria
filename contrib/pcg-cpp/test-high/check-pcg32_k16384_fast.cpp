#define RNG pcg32_k16384_fast
#define TWO_ARG_INIT 0

#include "pcg-test.cpp"

