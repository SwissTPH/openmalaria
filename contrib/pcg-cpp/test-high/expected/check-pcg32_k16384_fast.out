pcg32_k16384_fast:
      -  result:      32-bit unsigned int
      -  period:      2^524352
      -  size:        65544 bytes

Round 1:
  32bit: 0x3308e4a7 0x8e821ae6 0x0a95201f 0xbfc13f62 0x392ac135 0x9a9a3aab
  Again: 0x3308e4a7 0x8e821ae6 0x0a95201f 0xbfc13f62 0x392ac135 0x9a9a3aab
  Coins: HHHTHHTHTTTHHHHHTHHHTHTHTHTTTHTTTTHHHTHTHTHTHTTHTHHTTHHTHHTHHHHTH
  Rolls: 6 2 2 3 5 2 5 6 5 3 1 4 6 5 6 4 2 1 6 4 1 1 1 6 2 2 2 5 1 1 2 1 4
   -->   rolling dice used 33 random numbers
  Cards: 7h 4c 8s 7s 4d 2h 8c Th As 9s Qh Qc 4h 9d 6d Ad 6c 3s Kh 9h Js 5c
	 Tc Ks Ac 7c 3h 2d 5s Qd 3d 2c Qs 2s 6h 8h Jd 5h Kd Ts Kc Jh 4s 7d
	 Ah 9c Td 8d 5d 6s Jc 3c

Round 2:
  32bit: 0xdadd159c 0xbc2b894c 0xd0fe7abb 0xd3c99cc7 0xeec433b7 0xeb7f9d46
  Again: 0xdadd159c 0xbc2b894c 0xd0fe7abb 0xd3c99cc7 0xeec433b7 0xeb7f9d46
  Coins: THTTHTTHTTTTTHTTHTHHTTHTTHHTHTTTTTHHTTTHHTHTHHHHHTHHTHTTHHHTHTHHH
  Rolls: 3 6 3 5 5 6 2 5 1 5 6 6 2 6 3 6 6 2 6 3 5 1 2 6 5 1 5 3 3 2 2 5 3
   -->   rolling dice used 33 random numbers
  Cards: 2s 7s Qd Ah 4c Ad 8s 5d 5h 3s Ts Kc Js 5s 8c 9h 2d Kd 4h 9d Jh 3d
	 2c Tc Ac 3c 4s 5c 7h Qc Qh 6h As 8d 2h 6d 7c Kh 3h Jd Th 9s Jc 4d
	 Qs Td 7d 9c 6s 8h 6c Ks

Round 3:
  32bit: 0x1f019ace 0x4286124a 0x3f6959fc 0x6882a83f 0xd2c77978 0x721f00b2
  Again: 0x1f019ace 0x4286124a 0x3f6959fc 0x6882a83f 0xd2c77978 0x721f00b2
  Coins: HTHTHHTHTHHHTTTTTHHTTTHTTTHTTTHHTTHTHHHHTTHHHTTHHTHTHTTTHTHTTTHTH
  Rolls: 6 1 2 6 1 5 4 6 2 1 5 2 3 1 1 6 1 2 2 3 6 1 1 1 5 5 2 1 3 4 5 4 3
   -->   rolling dice used 33 random numbers
  Cards: 6s 4d 2d 7h 9c 4c 4h As Js 3d Ts 3s Jh 8h Qc Qd Ac 8s Jc Ks 3c 5c
	 5d Jd 7c 7s Ad 6c 6h Kd 9h Tc 2s 2h 5s 6d 7d 2c 8c Qh Kh 8d Td 5h
	 Qs Th Ah 4s 9s 9d Kc 3h

Round 4:
  32bit: 0xd607142f 0x2a130ecc 0xe1cdd7c8 0xbd5fd344 0xf003d169 0x0e7823df
  Again: 0xd607142f 0x2a130ecc 0xe1cdd7c8 0xbd5fd344 0xf003d169 0x0e7823df
  Coins: HTTHHHHTTHTHTHHTHHHHHTTTTHTHHHTHHHTHHHTTHTTHTHHTTTTTHHHHHHHTHHTHH
  Rolls: 1 4 3 2 4 3 2 5 1 4 4 4 3 1 4 3 3 2 5 3 1 2 5 6 4 3 5 1 2 1 6 1 6
   -->   rolling dice used 33 random numbers
  Cards: 8s 7c 7s Kh Ks 7h Tc 7d 9c Kc 3d 6c 4d 9d 5h 6s 3c 3s Ad Jd Qh Qd
	 4c Jh 6h Ah 8d 3h 5s 5c 4h Ts 4s 8c Td 2h Jc 2d 5d 2s 6d Th Ac 2c
	 9h 9s Qc Js Kd As Qs 8h

Round 5:
  32bit: 0x1511e2e9 0x51bde329 0x7a8b724b 0x60a82f6f 0x13bf7f80 0x97870401
  Again: 0x1511e2e9 0x51bde329 0x7a8b724b 0x60a82f6f 0x13bf7f80 0x97870401
  Coins: HTTTHHTTHHHTTHHTHTTTHTHHTTTHHTTHHHHHHTTTHHTHHHHHTHTHHTTHHTTHTHHTH
  Rolls: 5 3 4 4 5 6 4 6 6 4 4 4 5 5 6 2 4 2 6 4 5 3 2 6 1 5 3 2 4 3 5 2 3
   -->   rolling dice used 33 random numbers
  Cards: 8s As 2c Ad 7s Ac Kh Qh 5d 8c Jd Qc 8d 6h 5s 2h 2d 7c 7h Qs Jh 6c
	 Kd 3s Js 9d 4s 3c 8h 7d Qd Th 4c 3h 6s 9c Ah 5h 4d Kc Tc 9h Td 6d
	 3d 9s 4h 2s Ks Ts 5c Jc

